lumicks/pylake/recording.py

```diff
@@ -0,0 +1,154 @@
+        self.start_idx = start_idx
+        self.stop_idx = (self.src.num_frames if stop_idx is None else stop_idx)
+
+    def __getitem__(self, item):
+        """All indexing is in frames"""
+        if isinstance(item, slice):
+            if item.step is not None:
+                raise IndexError("Slice steps are not supported")
+
+            start, stop, _ = item.indices(self.num_frames)
+            return Recording(self.src, self.name, self.start_idx + start, self.start_idx + stop)
+        else:
+            item = self.start_idx + item if item >= 0 else self.stop_idx + item
+            if item >= self.stop_idx or item < self.start_idx:
+                raise IndexError("Index out of bounds")
+            return Recording(self.src, self.name, item, item + 1)
+
+    @staticmethod
+    def from_file(image_name):
+        """Construct recording from file
+
+        Parameters
+        ----------
+        image_name : str
+            Filename for the image stack.
+        """
+        return Recording(TiffStack.from_file(image_name), os.path.splitext(os.path.basename(image_name))[0])
+
+    def plot(self,  **kwargs):
+        import matplotlib.pyplot as plt
+
+        default_kwargs = dict(
+            cmap='gray'
+        )
+
+        frame = np.clip(kwargs.pop("frame", 0), 0, self.num_frames)
+        image = self.get_frame(frame).data
+        plt.imshow(image, **{**default_kwargs, **kwargs})
+
+        if self.num_frames == 1:
+            plt.title(self.name)
+        else:
+            plt.title(f"{self.name} [frame {frame}/{self.num_frames}]")
+
+    def get_frame(self, frame=0):
+        if frame >= self.num_frames or frame < 0:
+            raise IndexError("Frame index out of range")
+        return self.src.get_frame(self.start_idx + frame)
+
+    def downsample_channel(self, channel_slice, reduce=np.mean):
+        """Downsample channel on a frame by frame basis. The downsampling function (e.g. np.mean) is evaluated for the
+        time between a start and end time of a frame. A list is returned that contains the data corresponding to each
+        frame"
+
+        Parameters
+        ----------
+        channel_slice : pylake.channel.Slice
+            Data slice that we with to downsample.
+        reduce : callable
+            The `numpy` function which is going to reduce multiple samples into one.
+            The default is `np.mean`, but `np.sum` could also be appropriate for some
+            cases, e.g. photon counts.
+        """
+        t = np.zeros(self.num_frames)
+        d = np.zeros(self.num_frames)
+        for i, img_idx in enumerate(np.arange(self.start_idx, self.stop_idx)):
+            start, stop = (self.src.get_frame(img_idx).start, self.src.get_frame(img_idx).stop)
+            subset = channel_slice[start:stop]
+            t[i] = (start + stop) // 2
+            d[i] = reduce(subset.data)
+
+        return Slice(TimeSeries(d, t))
+
+    @property
+    def time_slice(self):
```

Comment:
There is a start and stop. Time_slice was removed because it was risky in case of dead times. Instead it has been replaced by slices so that you can directly iterate over the time range and get the slice data. I think this makes for a pretty nice syntax actually:

You can also obtain the force data corresponding to each frame directly:
``[h5.force1x[x.time_slice] for x in recording]``

Which I think is pretty nice. Start and stop are also in there so maybe it is indeed superfluous. You can also do ``[h5.force1x[x.start:x.stop] for x in recording]``

I'll remove it.

Reply:
Let's maybe leave this as a separate issue. I agree that we should have a terse syntax for indexing based on items. I'm not completely sold on `.time_slice`. I'd rather see indexing just by the item itself (see below). In any case I think this should be applied consistently to all objects rather than just this one.
```python
channel[item.start:item.stop] # verbose
channel[item.time_slice] # terse syntax option 1
channel[item] # terse syntax option 2
```

Reply:
Agreed. I've removed .time_slice. I'll add an improvement ticket for terse option #2, meant for adding it everywhere at the same time.